### ISSUE TITLE
random extremely slow response times of dockerd

### ISSUE THREAD

**Issue by piontec** [docker/docker#29635]
**Description**

We're running docker 1.12.4 on AWS on x1.32xlarge instance with Ubuntu 16.04. Randomly, the docker daemon responds very slowly to almost all API calls.

A live example caught when running `docker run`:
```
# docker run -it --rm ubuntu ipcs -l
docker: Error response from daemon: shim error: context deadline exceeded.
```

Logs from this time period:
```
Dec 21 11:09:58 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:09:58.589572735Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:09:59 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:09:59.498743699Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:23.123504197Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:23.124479766Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:23.125179984Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:25 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:25.924602801Z" level=error msg="containerd: start container" error="shim error: context deadline exceeded" id=14312312e55d35071264d211d2dfd8095454d78c62c39c44fb3e77a2e49f975a
Dec 21 11:10:25 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:25.925227778Z" level=error msg="Create container failed with error: shim error: context deadline exceeded"
Dec 21 11:10:26 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:26.347312214Z" level=error msg="Handler for POST /v1.24/containers/14312312e55d35071264d211d2dfd8095454d78c62c39c44fb3e77a2e49f975a/start returned error: shim error: context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.469826648Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.470011788Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.470005956Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.470051604Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.470715026Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:57.498872626Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:58 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:58.590113948Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:58 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:58.590108916Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:10:59 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:10:59.499523428Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:23.124304766Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:23.125017751Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:23 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:23.125686946Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:57.471164957Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:57.471337902Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:57.471380104Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
Dec 21 11:11:57 ip-10-69-11-89 dockerd[127631]: time="2016-12-21T11:11:57.471729462Z" level=warning msg="Health check error: rpc error: code = 4 desc = context deadline exceeded"
...
...
```

**Steps to reproduce the issue:**
1. Running dockerd with AUFS graphdriver, live-restore option enabled.

**Output of `docker version`:**

```
Client:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Tue Dec 13 00:08:34 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.4
 API version:  1.24
 Go version:   go1.6.4
 Git commit:   1564f02
 Built:        Tue Dec 13 00:08:34 2016
 OS/Arch:      linux/amd64

```

**Output of `docker info`:**

```
Containers: 443
 Running: 363
 Paused: 0
 Stopped: 80
Images: 1520
Server Version: 1.12.4
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 4340
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge null host overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-53-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8
```

**Additional info**
At that moment system had free CPU time (almost no io_wait) and over 50% of free memory.

**Comment by tonistiigi** [comment]
cc @mlaventure

**Comment by mlaventure** [comment]
@piontec what is the system doing? Is there a lot of docker commands being run in parallel? If yes, which one?

Is it possible to get the docker daemon logs in debug mode?

**Comment by piontec** [comment]
@mlaventure It's used solely as a docker host. It had about 400 containers running at the time and probably also multiple requests through TCP API endpoint. Local docker commands are rarely run.

It's a production host, so I can't restart the daemon at will, but I might try to enable debug log when I have a chance.

**Comment by Revolution1** [comment]
I scaled a service from 1 to 1000 replicas on a single-host cluster, and when the running task comes to 363, the number stopped increasing. 

```
# tail /var/log/upstart/docker.log
time="2016-12-22T15:04:35.570649270Z" level=error msg="fatal task error" error= module=taskmanager task.id=9fmjs1q98g6jm3nmvd29a7b0s
ERRO[0643] containerd: start container                   error=shim error: context deadline exceeded id=09e9455a3959e22ea908cb2a70c6651bc1ac47a2287bdc06b7d06ed522f5c702
time="2016-12-22T15:04:36.614903205Z" level=error msg="Create container failed with error: shim error: context deadline exceeded"
time="2016-12-22T15:04:36.816265788Z" level=error msg="fatal task error" error= module=taskmanager task.id=0kc4wh9so33wfdneonv8jbf9c
ERRO[0645] containerd: start container                   error=shim error: context deadline exceeded id=d6155809e601e375b642783e81bab80832243a2e2af49b41840c5fe7f7011578
time="2016-12-22T15:04:38.438621759Z" level=error msg="Create container failed with error: shim error: context deadline exceeded"
ERRO[0645] containerd: start container                   error=shim error: context deadline exceeded id=d6e7f2e14b78873d325967a5c0c1ff24877195e345732a1c442a3e91e624fd07
time="2016-12-22T15:04:38.484331548Z" level=error msg="Create container failed with error: shim error: context deadline exceeded"
time="2016-12-22T15:04:38.640796412Z" level=error msg="fatal task error" error= module=taskmanager task.id=8v67b7qega5mf6s8ugpqb2j2d
time="2016-12-22T15:04:38.745768686Z" level=error msg="fatal task error" error= module=taskmanager task.id=eet1ycxze3fvzg1v7bw1i4xeu
```

**Comment by RRAlex** [comment]
I have a similar issue where, after the fact, docker (1.12.5) stats only gives empty output, all the container fall in unhealthy mode, etc.
The system was running low on RAM, but even when something gets OOM-ed or memory is freed, it stays in that broken state. It seems like the Docker daemon can't recover from a low memory situation and is even timing out trying to kill containers or a simple `docker ps`.

```
CONTAINER           CPU %               MEM USAGE / LIMIT   MEM %               NET I/O             BLOCK I/O           PIDS
cbb96093e5d0        --                  -- / --             --                  -- / --             -- / --             --
fcfef32b9dea        --                  -- / --             --                  -- / --             -- / --             --
f096ca6c6729        --                  -- / --             --                  -- / --             -- / --             --
...
```

A reboot is basically required and it makes for a pretty useless `mem_limit:` if the containers don't get killed to free memory for the daemon to survive.

```
ERROR: An HTTP request took too long to complete. Retry with --verbose to obtain debug information.
If you encounter this issue regularly because of slow network conditions, consider setting COMPOSE_HTTP_TIMEOUT to a higher value (current value: 120).
```

Other interesting log lines (cherry-picked):
```
Dec 21 20:50:34 server0 dockerd[1380]: time="2016-12-21T20:50:34.489742521-05:00" level=warning msg="Health check error: context cancelled"
Dec 21 20:50:34 server0 dockerd[1380]: time="2016-12-21T20:50:34.489782799-05:00" level=warning msg="Health check error: context cancelled"
Dec 22 01:39:04 server0 dockerd[1380]: time="2016-12-22T01:39:04.146478388-05:00" level=error msg="stream copy error: reading from a closed fifo\ngithub.com/tonistiigi/fifo.(*fifo).Read\n\t/usr/src/docker/vendor/src/github.com/tonistiigi/fifo/fifo.go:141\nbufio.(*Reader).fill\n\t/usr/local/go/src/bufio/bufio.go:97\nbufio.(*Reader).WriteTo\n\t/usr/local/go/src/bufio/bufio.go:471\nio.copyBuffer\n\t/usr/local/go/src/io/io.go:370\nio.Copy\n\t/usr/local/go/src/io/io.go:350\ngithub.com/docker/docker/pkg/pools.Copy\n\t/usr/src/docker/.gopath/src/github.com/docker/docker/pkg/pools/pools.go:64\ngithub.com/docker/docker/container/stream.(*Config).CopyToPipe.func1.1\n\t/usr/src/docker/.gopath/src/github.com/docker/docker/container/stream/streams.go:119\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1998"
Dec 23 10:26:49 server0 dockerd[1380]: time="2016-12-23T10:26:49.506267503-05:00" level=error msg="containerd: start container" error="fork/exec /usr/bin/docker-containerd-shim: cannot allocate memory" id=91fe76066cd4825794aa6c5480e7ce197dd1a1472ad8d4200e77a46f2aa0bb21
Dec 23 10:26:49 server0 dockerd[1380]: time="2016-12-23T10:26:49.518211618-05:00" level=error msg="libcontainerd: error restarting rpc error: code = 2 desc = fork/exec /usr/bin/docker-containerd-shim: cannot allocate memory"
Dec 23 10:26:49 server0 dockerd[1380]: time="2016-12-23T10:26:49.519956311-05:00" level=error msg="libcontainerd: rpc error: code = 2 desc = fork/exec /usr/bin/docker-containerd-shim: cannot allocate memory"
Dec 23 10:29:08 server0 dockerd[1380]: time="2016-12-23T10:28:52.155014792-05:00" level=warning msg="Connect failed: dial udp 8.8.8.8:53: i/o timeout"
```

**Comment by mlaventure** [comment]
@Revolution1 your issue seems to be #29369

@RRAlex does restarting the stat command fix the issue after enough memory have been freed to allow a new program to be executed?

**Comment by RRAlex** [comment]
@mlaventure no, I really had to restart the daemon to get the stats back.

**Comment by mlaventure** [comment]
@RRAlex I'd say your issue is different than the one initially reported here, could you please open a new issue with all the details asked by the template?

**Comment by mlaventure** [comment]
@piontec could you set the daemon in debug mode and provide the logs there onwards?

You can do so without restarting the daemon by adding `debug: true` in `/etc/docker/daemon.json` and sending a `SIGHUP` to the daemon.

Thanks

**Comment by hitthecode** [comment]
The only way that this clears for me is to restart the daemon.

>Creating admin_queue_reservation_unit_notification_1

> 
> ERROR: for queue_reservation_unit_notification  Cannot start service queue_reservation_unit_notification: shim error: context deadline exceeded
> 
> ERROR: for queue_hkemail  Cannot start service queue_hkemail: shim error: context deadline exceeded
> 
> ERROR: for queue_unit_device_sync  Cannot start service queue_unit_device_sync: shim error: context deadline exceeded
> 
> ERROR: for queue_reservation_confirmation_email  Cannot start service queue_reservation_confirmation_email: shim error: context deadline exceeded
> 
> ERROR: for queue_photoupload  Cannot start service queue_photoupload: shim error: context deadline exceeded
> 
> ERROR: for queue_photodownload  Cannot start service queue_photodownload: shim error: context deadline exceeded
> 
> ERROR: for queue_command  Cannot start service queue_command: shim error: context deadline exceeded
> 
> ERROR: for queue_inbound_email  Cannot start service queue_inbound_email: shim error: context deadline exceeded
> ERROR: Encountered errors while bringing up the project.

**Comment by piontec** [comment]
I enabled debug log, but can't trigger this directly, when I'm running commands. Still, from monitoring, it seems that "docker ps" command is taking sometimes more than 5s and sometimes it gets stuck "for good" - the only solution is to restart dockerd.
If you let me know what to look for in the log, I will gladly search for it. I can also make the whole log available, but with debug we get about 1GB/day.

**Comment by mlaventure** [comment]
@piontec I forgot to ask last time, but before you restart `dockerd` next time, could you send a `SIGUSR1` to it to have it dump its stacktrace to the log? May help in finding the cause of the lock

**Comment by piontec** [comment]
@mlaventure I'm not sure if this is relevant, but currently just running `docker info` takes over 5s there. Maybe this can be a starting point to profiling? (I can easily repeat this):
```
# time docker info
Containers: 411
 Running: 323
 Paused: 0
 Stopped: 88
Images: 11391
Server Version: 1.12.6-cs7
Storage Driver: aufs
 Root Dir: /opt/io1/docker/aufs
 Backing Filesystem: extfs
 Dirs: 11188
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: apparmor seccomp
Kernel Version: 4.4.0-59-generic
Operating System: Ubuntu 16.04.1 LTS
OSType: linux
Architecture: x86_64
CPUs: 128
Total Memory: 1.876 TiB
Name: ip-10-69-11-89
ID: UGZS:UFD3:GB4C:W5MX:JU2L:K7PH:6ZWS:4GPM:27Q5:UNNN:X3DC:YDT7
Docker Root Dir: /opt/io1/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 3236
 Goroutines: 1983
 System Time: 2017-01-30T12:47:45.144432207Z
 EventsListeners: 1
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
Insecure Registries:
 127.0.0.0/8

real    0m5.485s
user    0m0.004s
sys     0m0.028s
```

**Comment by mlaventure** [comment]
@piontec can we get the stacktrace  of your daemon? You have close to 2000 goroutines running, I would have expected something closer to 1500 given the amount of containers you have running.

`docker info` have to go through all the containers to display the info, but I wouldn't expect it to take so long.

**Comment by panga** [comment]
I have a similar issue that always occurs with my tests after adding HEALTHCHECK to containers.
The issue happens in Docker 1.12.6+ , if I downgrade to 1.12.3 it works fine.

**Comment by panga** [comment]
@mlaventure 

**docker info**
```
Containers: 14
 Running: 13
 Paused: 0
 Stopped: 1
Images: 521
Server Version: 1.12.6
Storage Driver: overlay2
 Backing Filesystem: extfs
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: null host overlay bridge
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options:
Kernel Version: 4.4.44-39.55.amzn1.x86_64
Operating System: Amazon Linux AMI 2016.09
OSType: linux
Architecture: x86_64
CPUs: 2
Total Memory: 7.8 GiB
Name: ip-172-30-93-235
ID: SZVC:7CJG:HNCI:77MI:7DVU:X55J:G7HZ:NL4C:3CWR:KP47:NVFY:AKH3
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Insecure Registries:
 127.0.0.0/8
```

**log + stacktrace dump**
https://gist.github.com/panga/ac8c34402760430f2356ed8c3af9c4a0

**Comment by cpuguy83** [comment]
@panga Is it slow or stuck? Looks to me like it's stuck while waiting on a signal back from containerd.

<details>

```
goroutine 159173 [select, 3 minutes]:
google.golang.org/grpc/transport.(*Stream).Header(0xc822a34ee0, 0xc822a34ee0, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/vendor/src/google.golang.org/grpc/transport/transport.go:221 +0x242
google.golang.org/grpc.recvResponse(0x7fd4fac96b20, 0x2aa78a0, 0x0, 0x0, 0x0, 0x0, 0x7fd4fac96b58, 0xc8203e89c0, 0x7fd4fac96b80, 0xc8203e89e0, ...)
	/builddir/build/BUILD/docker-1.12.6/vendor/src/google.golang.org/grpc/call.go:54 +0x4e
google.golang.org/grpc.Invoke(0x7fd4fac88ae0, 0xc82005c5d8, 0x1e82790, 0x11, 0x1b565c0, 0xc8236d06f0, 0x1b56700, 0x2aa78a0, 0xc82039b6b0, 0x0, ...)
	/builddir/build/BUILD/docker-1.12.6/vendor/src/google.golang.org/grpc/call.go:178 +0xae6
github.com/docker/containerd/api/grpc/types.(*aPIClient).Signal(0xc820080620, 0x7fd4fac88ae0, 0xc82005c5d8, 0xc8236d06f0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/vendor/src/github.com/docker/containerd/api/grpc/types/api.pb.go:945 +0xec
github.com/docker/docker/libcontainerd.(*client).Signal(0xc8203f2c00, 0xc82224fb00, 0x40, 0xf, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/libcontainerd/client_linux.go:209 +0x19c
github.com/docker/docker/daemon.(*Daemon).kill(0xc8204261a0, 0xc8210b4000, 0xf, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/daemon/kill.go:156 +0x7c
github.com/docker/docker/daemon.(*Daemon).killWithSignal(0xc8204261a0, 0xc8210b4000, 0xf, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/daemon/kill.go:84 +0x491
github.com/docker/docker/daemon.(*Daemon).killPossiblyDeadProcess(0xc8204261a0, 0xc8210b4000, 0xf, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/daemon/kill.go:146 +0x5c
github.com/docker/docker/daemon.(*Daemon).containerStop(0xc8204261a0, 0xc8210b4000, 0xa, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/daemon/stop.go:48 +0xf9
github.com/docker/docker/daemon.(*Daemon).ContainerStop(0xc8204261a0, 0xc823646477, 0x40, 0xa, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/daemon/stop.go:28 +0x2bf
github.com/docker/docker/api/server/router/container.(*containerRouter).postContainersStop(0xc820224200, 0x7fd4facc01b8, 0xc8236d0600, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00, 0xc8236d0390, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/router/container/container_routes.go:168 +0x18b
github.com/docker/docker/api/server/router/container.(*containerRouter).(github.com/docker/docker/api/server/router/container.postContainersStop)-fm(0x7fd4facc01b8, 0xc8236d0600, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00, 0xc8236d0390, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/router/container/container.go:61 +0x74
github.com/docker/docker/api/server/middleware.VersionMiddleware.WrapHandler.func1(0x7fd4facc01b8, 0xc8236d0600, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00, 0xc8236d0390, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/middleware/version.go:56 +0x9aa
github.com/docker/docker/api/server/middleware.UserAgentMiddleware.WrapHandler.func1(0x7fd4facc01b8, 0xc8236d0570, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00, 0xc8236d0390, 0x0, 0x0)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/middleware/user_agent.go:45 +0x68e
github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/server.go:139 +0x138
net/http.HandlerFunc.ServeHTTP(0xc820199e60, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00)
	/usr/lib/golang/src/net/http/server.go:1618 +0x3a
github.com/gorilla/mux.(*Router).ServeHTTP(0xc82025e550, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00)
	/builddir/build/BUILD/docker-1.12.6/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e
github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc82075aa80, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00)
	/builddir/build/BUILD/docker-1.12.6/_build/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72
net/http.serverHandler.ServeHTTP(0xc8202b7f80, 0x7fd4fac5f848, 0xc820caf450, 0xc822a34e00)
	/usr/lib/golang/src/net/http/server.go:2081 +0x19e
net/http.(*conn).serve(0xc82287d580)
	/usr/lib/golang/src/net/http/server.go:1472 +0xf2e
created by net/http.(*Server).Serve
	/usr/lib/golang/src/net/http/server.go:2137 +0x44e
```

</details>

**Comment by panga** [comment]
@cpuguy83 , docker info is slow, docker ps stuck

**Comment by mlaventure** [comment]
@panga 1.12.6 has an issue where the context cancellation for exec/start command was not handled by containerd. This [PR](https://github.com/docker/containerd/pull/336/files) fixed it, but it was done after 1.12.6 was released.

This can be triggered when the machine is busy and cause the timeout of the healtcheck to be triggered before `containerd` has the time to process the `exec`

**Comment by panga** [comment]
@mlaventure Which docker version is linked to containerd that contains this fix?

**Comment by mlaventure** [comment]
@panga 1.13.0 onwards

**Comment by piontec** [comment]
@mlaventure Any progress on this? We're constantly monitoring `docker ps`, `docker info` and `docker run` response times now - see the attached graph. The case is that these response times are usually fairly constant until dockerd stops to reply at all. In that case, the only solution is to restart dockerd - then it gets back to normal. It was the case on the right side of graph.
![response_times](https://cloud.githubusercontent.com/assets/1578786/23701986/97524fea-03f9-11e7-8f77-1cfbccc3f176.png)

**Comment by mlaventure** [comment]
Closing this since it's resolved in a more recent release.

**Comment by matiaslespiau** [comment]
hi @mlaventure, which release fixes this issue?

**Comment by mlaventure** [comment]
@matiaslespiau any version after 1.13.0

**Comment by lwalid** [comment]
hi @mlaventure

i use docker 1.12.6, and i think that i have the seem problem. 
I have no answer for docker ps command or docker-compose ps command.
in the log i have :+1: 
Mar 27 12:04:02 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:02.078649648+02:00" level=info msg="containerd: 6c2cac230b04004330269d216097cd6ef71e98ccd5e94f3a8f874b3dc01bd358:8f4ce69aa39bc15Mar 27 12:04:02 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:02.399263983+02:00" level=warning msg="containerd: unable to save 78a48666f840e44d7796cf78601b9b7f99e7997e4a42a53e8e81fe6150f53dMar 27 12:04:02 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:02.39996041+02:00" level=info msg="containerd: 78a48666f840e44d7796cf78601b9b7f99e7997e4a42a53e8e81fe6150f53d78:2b8c824af751271fMar 27 12:04:04 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:04.961256646+02:00" level=warning msg="containerd: unable to save 36a96243cc84c164a4f47d059f1fb85fa0bb6401cd503d25bc8ba426faf54bMar 27 12:04:04 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:04.963955148+02:00" level=info msg="containerd: 36a96243cc84c164a4f47d059f1fb85fa0bb6401cd503d25bc8ba426faf54b72:df934ab55d43487Mar 27 12:04:05 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:05.024922976+02:00" level=warning msg="containerd: unable to save 5f266c9ec92c227b7294f3be60608d0bc4f686223ee28d5205c38af60f9d62Mar 27 12:04:05 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:05.027292935+02:00" level=info msg="containerd: 5f266c9ec92c227b7294f3be60608d0bc4f686223ee28d5205c38af60f9d6254:42275d834c94756Mar 27 12:04:05 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:05.087993324+02:00" level=warning msg="containerd: unable to save fec11c03476cc3257a90693f4f6942108031313ce7371e2ae7b0ce867d83b8Mar 27 12:04:05 test-ran.wanesy.fr dockerd[1059]: time="2017-03-27T12:04:05.088689234+02:00" level=info msg="containerd: fec11c03476cc3257a90693f4f6942108031313ce7371e2ae7b0ce867d83b858:aaeed1db1b7f6f7

**Comment by KevinTHU** [comment]
@mlaventure btw, is the PR "Remove timeout on fifos opening #30140"?

**Comment by mlaventure** [comment]
@KevinTHU I think your question is incomplete ;-)

**Comment by KevinTHU** [comment]
@mlaventure sorry, I thought you know that the PR link in your replay above is disabled, just for a check to ask you :-)

**Comment by mlaventure** [comment]
@KevinTHU It wasn't a PR.

The fix is present in docker releases versioned 1.13.x and above (including all the docker-ce releases since they came after 1.13.x)

**Comment by yogo1212** [comment]
I'm pretty sure I just got this with `Docker version 20.10.18, build b40c2f6` on Rocky Linux 9.
The machine has 2 GB RAM. Processing incoming mails (including amavis) took ~5 minutes.
Regular messages about a health check failing (`health check for container` .. `error: context deadline exceeded`) + 2 lines of closed fifo (`level=error msg="stream copy error: reading from a closed fifo"`).
That was after I added a swap file for another 2 GB.

Compose down/up didn't solve the problem and I'm pretty sure I restarted docker at one time (not containerd, though).
The message went away with a reboot.
